tensorflow/python/kernel_tests/bias_op_deterministic_test.py

```diff
@@ -0,0 +1,127 @@
+from tensorflow.python.ops import array_ops
+from tensorflow.python.ops import gradients_impl
+from tensorflow.python.ops import nn_ops
+from tensorflow.python.platform import test
+
+class BiasAddDeterministicTest(bias_op_base.BiasAddTestBase):
+
+  def _make_shape_tuple(self, batch_size, channel_count, data_rank, data_dim,
+                        data_layout):
+    data_dims = data_rank * (data_dim,)
+    if data_layout == 'channels_first':
+      shape = (batch_size,) + (channel_count,) + data_dims
+    elif data_layout == 'channels_last':
+      shape = (batch_size,) + data_dims + (channel_count,)
+    else:
+      raise ValueError("Unknown data format")
+    return shape
+
+  def _data_format_from_data_layout(self, data_layout=None):
+    if data_layout == 'channels_first':
+      return 'NCHW'
+    elif data_layout == 'channels_last':
+      return 'NHWC'
+    else:
+      raise ValueError("Unknown data_layout")
+
+  def _random_data_op(self, shape, data_type):
+    return constant_op.constant(
+        2 * np.random.random_sample(shape) - 1, dtype=data_type)
+
+  def _random_ndarray(self, shape):
+    return 2 * np.random.random_sample(shape) - 1
+
+  def _assert_reproducible(self, operation, feed_dict={}):
+    with self.cached_session(force_gpu=True):
+      result_a = operation[0].eval(feed_dict=feed_dict)
+      result_b = operation[0].eval(feed_dict=feed_dict)
+      self.assertAllEqual(result_a, result_b)
+
+  def _testDeterministicGradientsCase(self, data_layout, data_rank, data_type):
+    seed = (hash(data_layout) % 256 +
+            hash(data_rank) % 256 +
+            hash(data_type) % 256)
+    np.random.seed(seed)
+    batch_size = 10
+    channel_count = 8
+    data_dim = 14
+    in_shape = self._make_shape_tuple(batch_size, channel_count, data_rank,
+                                      data_dim, data_layout)
+    bias_shape = (channel_count,)
+    out_shape = in_shape
+    in_op = self._random_data_op(in_shape, data_type)
+    bias_op = self._random_data_op(bias_shape, data_type)
+    data_format = self._data_format_from_data_layout(data_layout)
+    bias_add_op = nn_ops.bias_add(in_op, bias_op, data_format=data_format)
+    upstream_gradients = array_ops.placeholder(data_type, shape=out_shape,
+                                               name='upstream_gradients')
+    gradient_injector_op = bias_add_op * upstream_gradients
+    # The gradient function behaves as if grad_ys is multiplied by the op
+    # gradient result, not passing the upstram gradients through the op's
+    # gradient generation graph. This is the reason for using the
+    # gradient_injector_op
+    grad_ys = None
+    bias_gradients_op = gradients_impl.gradients(
+        gradient_injector_op, bias_op, grad_ys=grad_ys,
+        colocate_gradients_with_ops=True)
+    for i in range(5):
+      feed_dict = {upstream_gradients: self._random_ndarray(out_shape)}
+      self._assert_reproducible(bias_gradients_op, feed_dict=feed_dict)
```

Comment:
This test is failing during the merging, it seems the result is not deterministic, we're still investigating. Do you have any idea?

Reply:
I'm wondering if the test is failing when run using XLA and that the XLA kernel fusion is replacing part of all of the deterministic solution with a non-deterministic kernel.

Reply:
In `tensorflow/tensorflow.bzl` in the master branch, it looks like `xla_enable_strict_auto_jit` is ignored. I'm wondering if it's not being ignored in the CI setup that the test is failing in. I'm wondering if `xla_enable_strict_auto_jit = True` should be removed from `tensorflow/python/kernel_tests/BUILD` for `bias_op_deterministic_test`. As an experiment, do you think it's worth me pushing a commit to this PR branch to remove it, @aaroey ?

Reply:
I  have reproduced what appears to be the same issue in a slightly different context. Enabling auto-JIT does make that test fail. XLA is probably replacing the deterministic mini-graph of ops with the original non-deterministic op kernel. I'll push a commit with auto-JIT disabled in a minute.

Reply:
Hi @aaroey, I just disabled XLA JIT for this test in the most recent commit. Please will review and run tests again and see if that resolves the issue.